### ISSUE TITLE
video-js memory leak?

### ISSUE THREAD

**Issue by heartsg** [videojs/video.js#2686]
I am mainly testing 4.12.7, but 5.0.0 has the same issue.

The current status:

http://stackoverflow.com/questions/32980125/memory-leak-in-javascript-and-video-js-on-all-browsers-with-the-simplest-exampl

**Comment by heartsg** [comment]
Btw, if I add vjs.cache = {} at the end of player.dispose(), then there's no leak. So it is for sure that something remained in vjs.cache which holds context of player that causes this leak (observed from chrome memory profiles).

**Comment by heartsg** [comment]
With the example given in stackoverflow, I recorded all created components and disposed components.

My observation is that, for example, the "subtitles off" was created twice (one with component cache id 204, and one with component cache id 304), the 304 was removed successfully from cache, but 204 was not.

**Comment by heartsg** [comment]
After 1 day of debugging, there are 2 issues of 4.12.7 (I didn't look at 5.0.0, but I think it should be similar because it's similar issues).

1. For textTrackSettings, all the SELECT etc options are built from template, and not inserted as children_. However, video-js calls vjs.on for these elements, which in turn calls getData, so these elements are saved into vjs.cache.

During dispose, they are not disposed from vjs.cache because they are not in children_ of textTrackSettings.

2. For menus (subtitleButton, CaptionsButton, Chaptersbutton), when you call update, you will remove old menu out of the children tree.

var menu = this.createMenu();

  if (this.menu) {
    this.removeChild(this.menu);
  }

However, you didn't remove these menu and menu items from vjs.cache.


Therefore, because vjs.cache is not cleared properly, they hold handlers and callbacks, which in turn holds context to these callbacks, which in turn holds the original player.

This causes memory leak.

**Comment by heartsg** [comment]
After the above are solved, there are still leaks (much smaller now, but still have). Trying to look into it. Hope I have time to completely solve it.

**Comment by gkatsev** [comment]
Thanks for looking into this. It's very possible we aren't doing as good of a job as we can with disposing things (like the cache). If you can submit a PR (or multiple) for changes, that would be awesome.

**Comment by stevensanborn** [comment]
Yes im seeing quite a number  leaks after calling dispose . There are still references to the main dom element hanging around.

**Comment by gkatsev** [comment]
This hasn't really been looked at in a while but this week I spent some time looking at this. A bunch of retained elements and other leaks were fixed in #4745 and TextTrackSettings was completely written since. As far as I can tell in dev tools, we no longer retain any unexpected objects on player dispose. I think it's time to close this issue and if we find more issues, we should open new issues for them.
